File: scripts/replay_local_provider_timing.py
#!/usr/bin/env python3
"""Replay Hades incremental provider deltas and cancellation through a PTY."""

from __future__ import annotations

import argparse
import json
import shutil
import threading
import time
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from typing import Any

from replay_local_provider import (
    COLUMNS,
    DEFAULT_BINARY,
    ROWS,
    ReplayFailure,
    assert_clean_exit,
    clean_output,
    marker_present,
    send,
    spawn,
    stop_process,
    wait_for,
    wait_for_exit,
)


ROOT = Path(__file__).resolve().parents[1]
DEFAULT_CONTRACT = ROOT / "tests/fixtures/parity/OBS-0053-hades-stream-timing.json"
FIRST_DELTA = "HADES_DELAY_FIRST"
SECOND_DELTA = "HADES_DELAY_SECOND"


class DelayedProviderServer(ThreadingHTTPServer):
    daemon_threads = True
    allow_reuse_address = True

    def __init__(self) -> None:
        self.request_seen = threading.Event()
        self.first_sent = threading.Event()
        self.release_second = threading.Event()
        self.handler_done = threading.Event()
        self.connection_closed = threading.Event()
        self.records: list[dict[str, Any]] = []
        self.writes: list[dict[str, Any]] = []
        super().__init__(("127.0.0.1", 0), self._handler())

    def _handler(self) -> type[BaseHTTPRequestHandler]:
        owner = self

        class Handler(BaseHTTPRequestHandler):
            def log_message(self, *_args: object) -> None:
                return

            def do_POST(self) -> None:  # noqa: N802 - stdlib handler API
                length = int(self.headers.get("Content-Length", "0"))
                body = self.rfile.read(length)
                owner.records.append(
                    {
                        "path": self.path,
                        "content_type": self.headers.get("Content-Type", ""),
                        "authorization_present": self.headers.get("Authorization") is not None,
                        "body": json.loads(body.decode("utf-8")),
                    }
                )
                owner.request_seen.set()
                self.send_response(200)
                self.send_header("Content-Type", "text/event-stream")
                self.send_header("Cache-Control", "no-cache")
                self.send_header("Connection", "close")
                self.end_headers()
                try:
                    self._write(0, b'data: {"choices":[{"delta":{"role":"assistant"}}]}\n\n', "role")
                    self._write(1, f'data: {{"choices":[{{"delta":{{"content":"{FIRST_DELTA}"}}}}]}}\n\n'.encode(), FIRST_DELTA)
                    owner.first_sent.set()
                    owner.release_second.wait(timeout=5.0)
                    self._write(2, f'data: {{"choices":[{{"delta":{{"content":"{SECOND_DELTA}"}}}}]}}\n\n'.encode(), SECOND_DELTA)
                    self._write(3, b'data: {"choices":[{"delta":{},"finish_reason":"stop"}]}\n\n', "finish")
                    self._write(4, b"data: [DONE]\n\n", "DONE")
                except (BrokenPipeError, ConnectionResetError, OSError) as error:
                    owner.connection_closed.set()
                    owner.writes.append({"index": 99, "marker": "connection-closed", "sent": False, "error": type(error).__name__})
                finally:
                    owner.handler_done.set()

            def _write(self, index: int, payload: bytes, marker: str) -> None:
                self.wfile.write(payload)
                self.wfile.flush()
                owner.writes.append({"index": index, "marker": marker, "sent": True})

        return Handler


def start_server() -> tuple[DelayedProviderServer, threading.Thread]:
    server = DelayedProviderServer()
    thread = threading.Thread(target=server.serve_forever, name="hades-delayed-provider", daemon=True)
    thread.start()
    return server, thread


def finish_server(server: DelayedProviderServer, thread: threading.Thread) -> None:
    server.release_second.set()
    server.shutdown()
    server.server_close()
    thread.join(timeout=2.0)


def run_case(binary: Path, interrupt: bool, timeout: float) -> dict[str, Any]:
    case = "interrupt-before-completion" if interrupt else "delayed-delta-order"
    server, server_thread = start_server()
    pid, fd, _slave_path, home = spawn(binary, f"http://127.0.0.1:{server.server_port}/v1")
    output = bytearray()
    reaped = False
    try:
        wait_for(pid, fd, output, case, "startup", lambda text: "Hermes Agent" in text and "ready" in text, timeout)
        send(fd, b"stream timing probe\r")
        wait_for(pid, fd, output, case, "request", lambda _text: server.request_seen.is_set(), timeout)
        wait_for(pid, fd, output, case, "first-write", lambda _text: server.first_sent.is_set(), timeout)
        wait_for(
            pid,
            fd,
            output,
            case,
            "first-delta",
            lambda text: marker_present(text, FIRST_DELTA),
            timeout,
        )

        if interrupt:
            send(fd, b"\x03")
            wait_for(pid, fd, output, case, "interrupt", lambda text: marker_present(text, "interrupted"), timeout)
            time.sleep(0.1)
            server.release_second.set()
            wait_for(pid, fd, output, case, "connection-close", lambda _text: server.handler_done.is_set(), timeout)
            if marker_present(clean_output(bytes(output)), SECOND_DELTA):
                raise ReplayFailure(case, "late-delta", "the second provider delta rendered after Ctrl+C", bytes(output))
            if not server.connection_closed.is_set():
                raise ReplayFailure(case, "connection-close", "provider server did not observe the cancelled socket", bytes(output))
            send(fd, b"\x03")
        else:
            server.release_second.set()
            wait_for(
                pid,
                fd,
                output,
                case,
                "second-delta",
                lambda text: marker_present(text, SECOND_DELTA),
                timeout,
            )
            wait_for(pid, fd, output, case, "ready", lambda text: marker_present(text, "ready"), timeout)
            send(fd, b"\x03")

        status = wait_for_exit(pid, fd, output, timeout)
        reaped = True
        assert_clean_exit(status, case)
        raw = bytes(output)
        if b"\x1b[?1049l" not in raw:
            raise ReplayFailure(case, "cleanup", "alternate screen was not restored", raw)
        if len(server.records) != 1:
            raise ReplayFailure(case, "request", f"expected one request, got {len(server.records)}", raw)
        record = server.records[0]
        body = record["body"]
        if record["path"] != "/v1/chat/completions" or record["content_type"] != "application/json":
            raise ReplayFailure(case, "request", "request boundary was not preserved", raw)
        if record["authorization_present"]:
            raise ReplayFailure(case, "request", "unexpected authorization crossed the sanitized replay boundary", raw)
        if sorted(body) != ["max_tokens", "messages", "model", "stream", "stream_options", "tools"]:
            raise ReplayFailure(case, "request", f"unexpected body keys: {sorted(body)}", raw)
        if body["model"] != "palette-model" or body["stream"] is not True:
            raise ReplayFailure(case, "request", "model/stream markers were not preserved", raw)
        if [message["role"] for message in body["messages"]] != ["system", "user"]:
            raise ReplayFailure(case, "request", "message role boundary was not preserved", raw)
        return {
            "id": case,
            "status": "passed",
            "input": [
                {"kind": "text", "value": "sanitized prompt"},
                {"kind": "key", "value": "Enter"},
                {"kind": "key", "value": "Ctrl+C", "meaning": "interrupt before delayed second delta" if interrupt else "bounded cleanup after completion"},
                *([{"kind": "key", "value": "Ctrl+C", "meaning": "exit from ready/interrupted state"}] if interrupt else []),
            ],
            "request": {
                "method": "POST",
                "path": record["path"],
                "content_type": record["content_type"],
                "authorization_present": False,
                "body_keys": sorted(body),
                "model": body["model"],
                "stream": body["stream"],
                "message_roles": [message["role"] for message in body["messages"]],
                "tools_present": bool(body["tools"]),
            },
            "writes": server.writes,
            "visible_state": {
                "first_delta_visible": True,
                "second_delta_visible": not interrupt,
                "partial_response_preserved": interrupt,
                "interrupted_surface_observed": interrupt,
                "returned_to_ready": True,
                "provider_connection_closed": server.connection_closed.is_set() if interrupt else False,
                "clean_exit": True,
            },
        }
    finally:
        stop_process(pid, fd, reaped)
        finish_server(server, server_thread)
        shutil.rmtree(home, ignore_errors=True)


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--binary", type=Path, default=DEFAULT_BINARY)
    parser.add_argument("--contract", type=Path, default=DEFAULT_CONTRACT)
    parser.add_argument("--report", type=Path, default=None)
    parser.add_argument("--timeout", type=float, default=8.0)
    args = parser.parse_args()
    binary = args.binary.resolve()
    report: dict[str, Any] = {
        "schema_version": 1,
        "command": "replay-local-provider-timing",
        "binary": str(binary),
        "contract": str(args.contract.resolve()),
        "dimensions": {"columns": COLUMNS, "rows": ROWS, "emulator": "direct PTY"},
        "cases": [],
        "passed": False,
    }
    try:
        if not binary.is_file():
            raise ReplayFailure("report", "binary", f"binary not found: {binary}")
        report["cases"] = [run_case(binary, interrupt=False, timeout=args.timeout), run_case(binary, interrupt=True, timeout=args.timeout)]
        report["passed"] = True
    except (OSError, ReplayFailure, json.JSONDecodeError, KeyError, TypeError, ValueError) as error:
        report["failure"] = error.as_dict() if isinstance(error, ReplayFailure) else {"message": str(error)}
        status = 1
    else:
        status = 0
    serialized = json.dumps(report, indent=2, sort_keys=True) + "\n"
    print(serialized, end="")
    if args.report is not None:
        args.report.parent.mkdir(parents=True, exist_ok=True)
        args.report.write_text(serialized, encoding="utf-8")
    return status


if __name__ == "__main__":
    raise SystemExit(main())

File: crates/hades-dev/src/bin/replay_local_provider_timing.rs
//! `hades-dev replay-local-provider-timing` — Rust port of
//! `scripts/replay_local_provider_timing.py` (HAD-156).
//!
//! Replays incremental provider deltas and Ctrl+C cancellation through a
//! direct PTY: the delayed loopback server writes a first delta, HOLDS
//! until release, then writes the second delta + finish + DONE. Two cases:
//! delayed-delta-order (both deltas render in order, clean exit) and
//! interrupt-before-completion (Ctrl+C between deltas cancels the socket;
//! the second delta must never render, then a second Ctrl+C exits).

use std::fs;
use std::io::{Read, Write};
use std::net::{Shutdown, TcpListener, TcpStream};
use std::path::{Path, PathBuf};
use std::process::ExitCode;
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::mpsc;
use std::time::{Duration, Instant};

use hades_dev::pty::read_available;
use hades_dev::replay::{
    ExitStatus, ReplayChild, marker_present, spawn_with_env, try_wait, wait_for, wait_for_exit,
    wait_for_rendered,
};
use serde_json::{Value, json};

const COLUMNS: u16 = 120;
const ROWS: u16 = 40;
const FIRST_DELTA: &str = "HADES_DELAY_FIRST";
const SECOND_DELTA: &str = "HADES_DELAY_SECOND";

struct DelayedState {
    request_seen: AtomicBool,
    first_sent: AtomicBool,
    release_second: AtomicBool,
    handler_done: AtomicBool,
    connection_closed: AtomicBool,
    shutdown: AtomicBool,
    records: std::sync::Mutex<Vec<RequestRecord>>,
    writes: std::sync::Mutex<Vec<Value>>,
}

#[derive(Clone)]
struct RequestRecord {
    path: String,
    content_type: String,
    authorization_present: bool,
    body: Value,
}

struct DelayedProvider {
    port: u16,
    state: Arc<DelayedState>,
    handle: Option<std::thread::JoinHandle<()>>,
}

impl DelayedProvider {
    fn start() -> std::io::Result<Self> {
        let listener = TcpListener::bind(("127.0.0.1", 0))?;
        let port = listener.local_addr()?.port();
        let state = Arc::new(DelayedState {
            request_seen: AtomicBool::new(false),
            first_sent: AtomicBool::new(false),
            release_second: AtomicBool::new(false),
            handler_done: AtomicBool::new(false),
            connection_closed: AtomicBool::new(false),
            shutdown: AtomicBool::new(false),
            records: std::sync::Mutex::new(Vec::new()),
            writes: std::sync::Mutex::new(Vec::new()),
        });
        let thread_state = Arc::clone(&state);
        let handle = std::thread::spawn(move || serve(listener, thread_state));
        Ok(Self { port, state, handle: Some(handle) })
    }

    fn environment(&self) -> Vec<(&'static str, String)> {
        vec![
            ("HADES_PROVIDER_BASE_URL", format!("http://127.0.0.1:{}/v1", self.port)),
            ("HADES_MODEL", "palette-model".to_string()),
            ("HADES_PROVIDER_API_KEY", String::new()),
        ]
    }

    fn finish(&mut self) {
        self.state.release_second.store(true, Ordering::SeqCst);
        self.state.shutdown.store(true, Ordering::SeqCst);
        if let Some(handle) = self.handle.take() {
            let _ = handle.join();
        }
    }
}

fn parse_head(buf: &[u8]) -> Result<(usize, String, String, bool, usize), String> {
    let text = String::from_utf8_lossy(buf);
    let mut path = String::new();
    let mut content_type = String::new();
    let mut authorization_present = false;
    let mut content_length = 0usize;
    for line in text.split('\n') {
        let line = line.strip_suffix('\r').unwrap_or(line);
        if line.is_empty() {
            continue;
        }
        if path.is_empty() {
            let mut parts = line.split_whitespace();
            let _method = parts.next();
            if let Some(part) = parts.next() {
                path = part.to_string();
            }
            continue;
        }
        if let Some((key, value)) = line.split_once(':') {
            match key.trim().to_ascii_lowercase().as_str() {
                "content-length" => content_length = value.trim().parse().unwrap_or(0),
                "content-type" => content_type = value.trim().to_string(),
                "authorization" => authorization_present = true,
                _ => {}
            }
        }
    }
    let header_end = buf
        .windows(4)
        .position(|window| window == b"\r\n\r\n")
        .ok_or_else(|| "missing header boundary".to_owned())?;
    Ok((header_end, path, content_type, authorization_present, content_length))
}

fn read_request(stream: &mut TcpStream) -> Result<RequestRecord, String> {
    let mut buf = Vec::new();
    let mut tmp = [0u8; 8192];
    let head = loop {
        let count =
            stream.read(&mut tmp).map_err(|error| format!("request read failed: {error}"))?;
        if count == 0 {
            return Err("request ended inside the head".to_string());
        }
        buf.extend_from_slice(&tmp[..count]);
        if buf.windows(4).any(|window| window == b"\r\n\r\n") {
            break parse_head(&buf)?;
        }
    };
    let (header_end, path, content_type, authorization_present, content_length) = head;
    let mut body_bytes: Vec<u8> = buf[header_end + 4..].to_vec();
    while body_bytes.len() < content_length {
        let count =
            stream.read(&mut tmp).map_err(|error| format!("request body read failed: {error}"))?;
        if count == 0 {
            break;
        }
        body_bytes.extend_from_slice(&tmp[..count]);
    }
    body_bytes.truncate(content_length);
    let body = serde_json::from_slice(&body_bytes).unwrap_or(Value::Null);
    Ok(RequestRecord { path, content_type, authorization_present, body })
}

fn write_chunk(
    stream: &mut TcpStream,
    state: &DelayedState,
    index: usize,
    marker: &str,
    payload: &[u8],
) -> bool {
    let sent = stream.write_all(payload).is_ok() && stream.flush().is_ok();
    if !sent {
        // Mirror the Python handler: a failed write surfaces as the
        // connection-closed record (index 99), not the write itself.
        state.writes.lock().unwrap().push(json!({
            "index": 99,
            "marker": "connection-closed",
            "sent": false,
            "error": "BrokenPipeError",
        }));
        return false;
    }
    state.writes.lock().unwrap().push(json!({
        "index": index,
        "marker": marker,
        "sent": true,
    }));
    true
}

fn handle_connection(mut stream: TcpStream, state: Arc<DelayedState>) {
    let _ = stream.set_nodelay(true);
    let record = match read_request(&mut stream) {
        Ok(record) => record,
        Err(_) => return,
    };
    state.records.lock().unwrap().push(record);
    state.request_seen.store(true, Ordering::SeqCst);

    let head = b"HTTP/1.1 200 OK\r\nContent-Type: text/event-stream\r\nCache-Control: no-cache\r\nConnection: close\r\n\r\n";
    if stream.write_all(head).is_err() {
        state.connection_closed.store(true, Ordering::SeqCst);
        state.writes.lock().unwrap().push(json!({
            "index": 99,
            "marker": "connection-closed",
            "sent": false,
            "error": "write head failed",
        }));
        state.handler_done.store(true, Ordering::SeqCst);
        return;
    }
    write_chunk(
        &mut stream,
        &state,
        0,
        "role",
        b"data: {\"choices\":[{\"delta\":{\"role\":\"assistant\"}}]}\n\n",
    );
    let first =
        format!("data: {{\"choices\":[{{\"delta\":{{\"content\":\"{FIRST_DELTA}\"}}}}]}}\n\n");
    write_chunk(&mut stream, &state, 1, FIRST_DELTA, first.as_bytes());
    state.first_sent.store(true, Ordering::SeqCst);

    // Hold the second delta until release (bounded like the Python 5s).
    let deadline = Instant::now() + Duration::from_secs(5);
    while !state.release_second.load(Ordering::SeqCst) && Instant::now() < deadline {
        std::thread::sleep(Duration::from_millis(10));
    }

    let second =
        format!("data: {{\"choices\":[{{\"delta\":{{\"content\":\"{SECOND_DELTA}\"}}}}]}}\n\n");
    let ok = write_chunk(&mut stream, &state, 2, SECOND_DELTA, second.as_bytes())
        && write_chunk(
            &mut stream,
            &state,
            3,
            "finish",
            b"data: {\"choices\":[{\"delta\":{},\"finish_reason\":\"stop\"}]}\n\n",
        )
        && write_chunk(&mut stream, &state, 4, "DONE", b"data: [DONE]\n\n");
    if !ok {
        state.connection_closed.store(true, Ordering::SeqCst);
    }
    let _ = stream.shutdown(Shutdown::Both);
    state.handler_done.store(true, Ordering::SeqCst);
}

fn serve(listener: TcpListener, state: Arc<DelayedState>) {
    let _ = listener.set_nonblocking(true);
    loop {
        if state.shutdown.load(Ordering::SeqCst) {
            break;
        }
        match listener.accept() {
            Ok((stream, _)) => {
                let thread_state = Arc::clone(&state);
                std::thread::spawn(move || handle_connection(stream, thread_state));
            }
            Err(_) => std::thread::sleep(Duration::from_millis(10)),
        }
    }
}

fn spawn_tui(binary: &Path, base_url: &str) -> Result<ReplayChild, String> {
    let home = std::env::temp_dir().join(format!("hades-timing-{}", std::process::id()));
    fs::create_dir_all(&home).map_err(|error| error.to_string())?;
    let extra: [(&str, &str); 4] = [
        ("HADES_PROVIDER_BASE_URL", base_url),
        ("HADES_MODEL", "palette-model"),
        ("HOME", home.to_str().ok_or_else(|| "non-utf8 home".to_owned())?),
        ("HERMES_HOME", home.to_str().ok_or_else(|| "non-utf8 home".to_owned())?),
    ];
    let child = spawn_with_env(binary, &[], &extra, &["HADES_PROVIDER_API_KEY"])
        .map_err(|error| error.to_string())?;
    Ok(child)
}

fn assert_request_boundary(record: &RequestRecord) -> Result<(), String> {
    let body = &record.body;
    if record.path != "/v1/chat/completions" || record.content_type != "application/json" {
        return Err("request boundary was not preserved".to_owned());
    }
    if record.authorization_present {
        return Err("unexpected authorization crossed the sanitized replay boundary".to_owned());
    }
    let keys = body
        .as_object()
        .map(|map| {
            let mut keys: Vec<&str> = map.keys().map(String::as_str).collect();
            keys.sort();
            keys
        })
        .unwrap_or_default();
    if keys != ["max_tokens", "messages", "model", "stream", "stream_options", "tools"] {
        return Err(format!("unexpected body keys: {keys:?}"));
    }
    if body["model"] != "palette-model" || body["stream"] != true {
        return Err("model/stream markers were not preserved".to_owned());
    }
    let roles: Vec<&str> = body["messages"]
        .as_array()
        .map(|messages| {
            messages
                .iter()
                .filter_map(|message| message.get("role").and_then(Value::as_str))
                .collect()
        })
        .unwrap_or_default();
    if roles != ["system", "user"] {
        return Err("message role boundary was not preserved".to_owned());
    }
    Ok(())
}

fn run_case(binary: &Path, interrupt: bool, timeout: Duration) -> Result<Value, String> {
    let case = if interrupt { "interrupt-before-completion" } else { "delayed-delta-order" };
    let mut server = DelayedProvider::start().map_err(|error| error.to_string())?;
    let base_url = format!("http://127.0.0.1:{}/v1", server.port);
    let mut child = spawn_tui(binary, &base_url)?;
    let mut output = Vec::new();
    let mut reaped = false;

    let result = (|| -> Result<Value, String> {
        wait_for(
            &child.child,
            &mut output,
            &format!("{case}: startup"),
            |text| marker_present(text, "Hades Agent") && marker_present(text, "ready"),
            timeout,
        )?;
        hades_dev::replay::send(&child.child.master, b"stream timing probe\r")
            .map_err(|error| error.to_string())?;
        wait_for(
            &child.child,
            &mut output,
            &format!("{case}: request"),
            |_| server.state.request_seen.load(Ordering::SeqCst),
            timeout,
        )?;
        wait_for(
            &child.child,
            &mut output,
            &format!("{case}: first-write"),
            |_| server.state.first_sent.load(Ordering::SeqCst),
            timeout,
        )?;
        wait_for_rendered(
            &child.child,
            &mut output,
            &format!("{case}: first-delta"),
            |rendered| marker_present(rendered, FIRST_DELTA),
            timeout,
        )?;

        if interrupt {
            hades_dev::replay::send(&child.child.master, b"\x03")
                .map_err(|error| error.to_string())?;
            wait_for_rendered(
                &child.child,
                &mut output,
                &format!("{case}: interrupt"),
                |rendered| marker_present(rendered, "interrupted"),
                timeout,
            )?;
            std::thread::sleep(Duration::from_millis(100));
            server.state.release_second.store(true, Ordering::SeqCst);
            wait_for(
                &child.child,
                &mut output,
                &format!("{case}: connection-close"),
                |_| server.state.handler_done.load(Ordering::SeqCst),
                timeout,
            )?;
            let text = rendered_text(&output);
            if marker_present(&text, SECOND_DELTA) {
                return Err(format!(
                    "{case}: late-delta: the second provider delta rendered after Ctrl+C"
                ));
            }
            if !server.state.connection_closed.load(Ordering::SeqCst) {
                return Err(format!(
                    "{case}: connection-close: provider server did not observe the cancelled socket"
                ));
            }
            hades_dev::replay::send(&child.child.master, b"\x03")
                .map_err(|error| error.to_string())?;
        } else {
            server.state.release_second.store(true, Ordering::SeqCst);
            wait_for_rendered(
                &child.child,
                &mut output,
                &format!("{case}: second-delta"),
                |rendered| marker_present(rendered, SECOND_DELTA),
                timeout,
            )?;
            wait_for_rendered(
                &child.child,
                &mut output,
                &format!("{case}: ready"),
                |rendered| marker_present(rendered, "ready"),
                timeout,
            )?;
            hades_dev::replay::send(&child.child.master, b"\x03")
                .map_err(|error| error.to_string())?;
        }

        let status = wait_for_exit(&child.child, &mut output, timeout)?;
        reaped = true;
        let exit_status = ExitStatus::describe(status);
        if exit_status != (ExitStatus::Exit { code: 0 }) {
            return Err(format!("{case}: unexpected exit status: {exit_status:?}"));
        }
        if !find_sequence(output.as_slice(), b"\x1b[?1049l") {
            return Err(format!("{case}: cleanup: alternate screen was not restored"));
        }

        let records = server.state.records.lock().unwrap().clone();
        if records.len() != 1 {
            return Err(format!("{case}: request: expected one request, got {}", records.len()));
        }
        let record = &records[0];
        assert_request_boundary(record)?;
        let writes = server.state.writes.lock().unwrap().clone();
        let second_delta_visible = !interrupt;
        let mut input = vec![
            json!({"kind": "text", "value": "sanitized prompt"}),
            json!({"kind": "key", "value": "Enter"}),
            json!({
                "kind": "key",
                "value": "Ctrl+C",
                "meaning": if interrupt {
                    "interrupt before delayed second delta"
                } else {
                    "bounded cleanup after completion"
                },
            }),
        ];
        if interrupt {
            input.push(json!({
                "kind": "key",
                "value": "Ctrl+C",
                "meaning": "exit from ready/interrupted state",
            }));
        }
        Ok(json!({
            "id": case,
            "status": "passed",
            "input": input,
            "request": {
                "method": "POST",
                "path": record.path,
                "content_type": record.content_type,
                "authorization_present": false,
                "body_keys": record.body.as_object().map(|map| {
                    let mut keys: Vec<&str> = map.keys().map(String::as_str).collect();
                    keys.sort();
                    keys
                }).unwrap_or_default(),
                "model": record.body["model"],
                "stream": record.body["stream"],
                "message_roles": record.body["messages"].as_array().map(|messages| {
                    messages.iter().filter_map(|m| m.get("role").and_then(Value::as_str)).collect::<Vec<_>>()
                }).unwrap_or_default(),
                "tools_present": record.body["tools"].as_array().is_some_and(|tools| !tools.is_empty()),
            },
            "writes": writes,
            "visible_state": {
                "first_delta_visible": true,
                "second_delta_visible": second_delta_visible,
                "partial_response_preserved": interrupt,
                "interrupted_surface_observed": interrupt,
                "returned_to_ready": true,
                "provider_connection_closed": if interrupt {
                    server.state.connection_closed.load(Ordering::SeqCst)
                } else {
                    false
                },
                "clean_exit": true,
            },
        }))
    })();

    if !reaped {
        hades_dev::pty::stop(&mut child.child.child);
    }
    server.finish();
    let _ = fs::remove_dir_all(
        &std::env::temp_dir().join(format!("hades-timing-{}", std::process::id())),
    );
    result
}

fn rendered_text(output: &[u8]) -> String {
    let mut screen = hades_dev::screen::Screen::new(COLUMNS as usize, ROWS as usize);
    screen.feed(output);
    screen.lines().join("\n")
}

fn find_sequence(haystack: &[u8], needle: &[u8]) -> bool {
    haystack.windows(needle.len()).any(|window| window == needle)
}

fn write_report(report: &Value, path: Option<&Path>, status: u8) -> ExitCode {
    let mut text = serde_json::to_string_pretty(report).expect("serialize report");
    text.push('\n');
    if let Some(path) = path {
        if let Some(parent) = path.parent() {
            let _ = fs::create_dir_all(parent);
        }
        use std::io::Write;
        if let Ok(mut file) = fs::File::create(path) {
            let _ = file.write_all(text.as_bytes());
        }
    }
    print!("{text}");
    if status == 0 { ExitCode::SUCCESS } else { ExitCode::from(status) }
}

fn main() -> ExitCode {
    let mut args = std::env::args().skip(1);
    let mut binary = PathBuf::from("target/debug/hades");
    let mut contract_path =
        PathBuf::from("tests/fixtures/parity/OBS-0053-hades-stream-timing.json");
    let mut report_path: Option<PathBuf> = None;
    let mut timeout = Duration::from_secs_f64(8.0);

    while let Some(arg) = args.next() {
        match arg.as_str() {
            "--binary" => {
                if let Some(value) = args.next() {
                    binary = PathBuf::from(value);
                }
            }
            "--contract" => {
                if let Some(value) = args.next() {
                    contract_path = PathBuf::from(value);
                }
            }
            "--report" => {
                if let Some(value) = args.next() {
                    report_path = Some(PathBuf::from(value));
                }
            }
            "--timeout" => {
                if let Some(value) = args.next() {
                    timeout = Duration::from_secs_f64(value.parse().unwrap_or(8.0));
                }
            }
            _ => {}
        }
    }

    let binary = binary.canonicalize().unwrap_or_else(|_| binary.clone());
    let contract_path = contract_path.canonicalize().unwrap_or_else(|_| contract_path.clone());
    let mut report = json!({
        "schema_version": 1,
        "command": "replay-local-provider-timing",
        "binary": binary.display().to_string(),
        "contract": contract_path.display().to_string(),
        "dimensions": {"columns": COLUMNS, "rows": ROWS, "emulator": "direct PTY"},
        "cases": [],
        "passed": false,
    });
    if !binary.is_file() {
        report["failure"] = json!({"case": "report", "step": "binary", "message": format!("binary not found: {}", binary.display())});
        return write_report(&report, report_path.as_deref(), 1);
    }

    let mut cases = Vec::new();
    for interrupt in [false, true] {
        match run_case(&binary, interrupt, timeout) {
            Ok(case) => cases.push(case),
            Err(error) => {
                report["failure"] = json!({"case": "report", "step": "runtime", "message": error});
                return write_report(&report, report_path.as_deref(), 1);
            }
        }
    }
    report["cases"] = json!(cases);
    report["passed"] = json!(true);
    write_report(&report, report_path.as_deref(), 0)
}
